File: demo-task-executor-java/src/main/java/demo/executor/Task.java
package demo.executor;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.Callable;

/**
 * Task that combines {@link java.util.concurrent.Callable} as producer and {@link demo.executor.Action} as consumer.
 * This class IS thread-safe.
 */
final public class Task<T> {

    final private String name;
    final private Class<T> clazz;
    final private Callable<T> callable;
    final private Action<T> action;

    private volatile boolean finished = false;
    private LocalDateTime start = LocalDateTime.now();
    private Optional<Duration> duration = Optional.empty();

    /**
     * Creates an task.
     *
     * @param name     the task name.
     * @param clazz    the result class type token to cast generic type at runtime.
     * @param callable the {@link java.util.concurrent.Callable} to run.
     * @param action   the {@link demo.executor.Action} to process result or exception.
     */
    public Task(String name, Class<T> clazz, Callable<T> callable, Action<T> action) {
        this.name = Objects.requireNonNull(name);
        this.clazz = Objects.requireNonNull(clazz);
        this.callable = Objects.requireNonNull(callable);
        this.action = Objects.requireNonNull(action);
    }

    Callable<T> getCallable() {
        return this.callable;
    }

    boolean isFinished() {
        return finished;
    }

    void markFinished() {
        if (finished) {
            throw new IllegalStateException("Task must be non-finished");
        }
        duration = Optional.of(Duration.between(start, LocalDateTime.now()));
        finished = true;
    }

    synchronized void setResult(Object result) {
        this.action.accept(Optional.of(clazz.cast(result)), Optional.empty());
        markFinished();
    }

    synchronized void setException(Exception exception) {
        this.action.accept(Optional.empty(), Optional.of(Objects.requireNonNull(exception)));
        markFinished();
    }

    /**
     * Waits until the task is finished.
     *
     * @return the task duration.
     */
    public Duration waitUntilFinished() {
        while (!finished) ;
        if (!duration.isPresent()) {
            throw new IllegalStateException("Duration must be set");
        }
        return duration.get();
    }

    @Override
    public String toString() {
        return "Task{" + "'" + name + '\'' + '}';
    }
}

File: demo-task-executor-kotlin/src/main/kotlin/demo/executor/Task.kt
package demo.executor

import java.time.Duration
import java.time.LocalDateTime
import java.util.Optional
import java.util.concurrent.Callable
import kotlin.reflect.KClass

/**
 * Task that combines [java.util.concurrent.Callable] as producer and [demo.executor.Action] as consumer.
 * This class IS thread-safe.
 */
class Task<T : Any>
/**
 * Creates an task.

 * @param name     the task name.
 * @param clazz    the result class type token to cast generic type at runtime.
 * @param callable the [java.util.concurrent.Callable] to run.
 * @param action   the action to process result or exception.
 */
(name: String, clazz: KClass<T>, callable: Callable<T>, action: (Optional<T>, Optional<Exception>) -> Unit) {

    private val name: String
    private val clazz: KClass<T>
    private val callable: Callable<T>
    private val action: (Optional<T>, Optional<Exception>) -> Unit

    @Volatile internal var finished = false
    private val start = LocalDateTime.now()
    private var duration = Optional.empty<Duration>()

    init {
        this.name = requireNotNull(name)
        this.clazz = requireNotNull(clazz)
        this.callable = requireNotNull(callable)
        this.action = requireNotNull(action)
    }

    internal fun getCallable(): Callable<T> {
        return this.callable
    }

    internal fun isFinished(): Boolean {
        return finished
    }

    internal fun markFinished() {
        check (!finished, { "Task must be non-finished" })
        duration = Optional.of(Duration.between(start, LocalDateTime.now()))
        finished = true
    }

    @Synchronized internal fun setResult(result: Any) {
        val result2 = if (result is Number) clazz.javaObjectType.cast(result) else clazz.java.cast(result)
        this.action(Optional.of(result2), Optional.empty<Exception>())
        markFinished()
    }

    @Synchronized internal fun setException(exception: Exception) {
        this.action(Optional.empty<T>(), Optional.of(requireNotNull(exception)))
        markFinished()
    }

    /**
     * Waits until the task is finished.

     * @return the task duration.
     */
    fun waitUntilFinished(): Duration {
        while (!finished) {
        }
        require (duration.isPresent, { "Duration must be set" })
        return duration.get()
    }

    override fun toString(): String {
        return "Task{'$name'}"
    }
}
